src/_pytest/python.py

```diff
@@ -747,7 +747,10 @@ def collect(self) -> Iterable[Union[nodes.Item, nodes.Collector]]:
         this_path = self.path.parent
 
         # Always collect the __init__ first.
-        if path_matches_patterns(self.path, self.config.getini("python_files")):
+        if (
+            self.session.isinitpath(self.path)
```

Comment:
`self.path` here is the `__init__.py` file (== the `path` of the `Package`).

This says, if `pkg/__init__.py` was specified directly (e.g. `pytest pkg/__init__.py`), then collect its `Module` even if `__init__.py` is not a recognized pattern in `python_files` config.

The relevance of this to the bug is the following code in `Session`:

https://github.com/pytest-dev/pytest/blob/4f3f36c396b52f8398bc4734ff0c00c57cf1fed1/src/_pytest/main.py#L805-L818

That code is currently broken.